blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc

```diff
@@ -0,0 +1,455 @@
+= Materializing Aggregate Views With Hibernate and Debezium
+gmorling
+:awestruct-tags: [ discussion, examples ]
+:awestruct-layout: blog-post
+
+Updating external full text search indexes (e.g. https://www.elastic.co/products/elasticsearch[Elasticsearch]) after data changes is a very popular use case for change data capture (CDC).
+
+As we've discussed in a link:/blog/2018/01/17/streaming-to-elasticsearch/[blog post] a while ago,
+the combination of Debezium's CDC source connectors and Confluent's https://docs.confluent.io/current/connect/connect-elasticsearch/docs/index.html[sink connector for Elasticsearch] makes it straight forward to capture data changes in MySQL, Postgres etc. and push them towards Elastisearch in near real-time.
+This results in a 1:1 relationship between tables in the source database and a corresponding search index in Elasticsearch,
+which is perfectly fine for many use cases.
+
+It gets more challenging though if you'd like to put entire aggregates into a single index.
+An example could be a customer and all their addresses;
+those would typically be stored in two separate tables in an RDBMS, linked by a foreign key,
+whereas you'd like to have just one index in Elasticsearch,
+containing documents of customers with their addresses embedded,
+allowing you to efficiently search for customers based on their address.
+
+Following up to the link:/blog/2018/03/08/creating-ddd-aggregates-with-debezium-and-kafka-streams/[KStreams-based solution] to this we described recently,
+we'd like to present in this post an alternative for materializing such aggregate views driven by the application layer.
+
+== Overview
+
+The idea is to materialize views in a separate table in the source database,
+right in the moment the primary data is altered.
+
+Aggregates are serialized as JSON structure (which naturally can represent any nested object structure) and stored in a specific table.
+This is done within the actual transaction altering the data,
+which means the aggregate view is always consistent with the primary data.
+In particular this approach isn't prone to exposing intermediary aggregations as the KStreams-based solution discussed in the post linked above.
+
+The following picture shows the overall architecture:
+
+[.centered-image.responsive-image]
+++++
+<img src="/images/jpa_aggregations.png" style="max-width:100%; margin-bottom:10px;" class="responsive-image" alt="Streaming Materialized Aggregate Views to Elastisearch">
+++++
+
+Here the aggregate views are materialized by means of a small extension to http://hibernate.org/orm/[Hibernate ORM],
```

Comment:
while conceptually comparable to a materialized view it's strictly speaking not such a thing we are dealing with here. at the end it's building aggregates which are written into a "normal table" or did I maybe misunderstand the approach taken?

Reply:
I agree I'm not 100% consistent with the terms and switch a bit forth and back. To me it definitely is a materialized view, though, just not in the sense of a SQL materialized view. I'd argue though that also in an RDBMS without native mat view support you could create materialized views yourself and store them in a table. And in a way that's what we're doing here.

Reply:
I've added a sentence to clarify things a bit.